tinybytes/src/lib.rs

```diff
@@ -154,38 +149,49 @@ impl Bytes {
 
         let subset_start = subset.as_ptr() as usize;
         let subset_end = subset_start + subset.len();
-        let self_start = self.ptr as usize;
-        let self_end = self_start + self.len;
+        let self_start = self.slice.as_ptr() as usize;
+        let self_end = self_start + self.slice.len();
         if subset_start >= self_start && subset_end <= self_end {
             Some(self.safe_slice_ref(subset_start - self_start, subset_end - self_start))
         } else {
             None
         }
     }
 
+    /// Returns a mutable reference to the slice of self.
+    /// Allows for fast unchecked shrinking of the slice.
+    ///
+    /// # Safety
+    ///
+    /// Callers of that function must make sure that they only put subslices of the slice into the
+    /// returned reference.
+    /// They also need to make sure to not persist the slice reference for longer than the struct
+    /// lives.
+    #[inline]
+    pub unsafe fn as_mut_slice(&mut self) -> &mut &'static [u8] {
+        &mut self.slice
+    }
```

Comment:
It's not just as simple as that, In particular with the string reading:
```
    read_string_ref_nomut(unsafe { buf.as_mut_slice() }).map(|(str, newbuf)| {
        let string = BytesString::from_bytes_slice(buf, str);
        *unsafe { buf.as_mut_slice() } = newbuf;
        string
    })
```
there's a dance around the ordering of executions: _first_, read the string. Then fetch the string slice. Then adjust the size.
Because `Bytes` asserts that any read from it is within its bounds. So we'd then have to craft an unsafe method for _that_ instead.

Also, your `Read` impl is actually copying every single byte. Passing a slice directly does not incur that overhead. It's maybe a bit more elegant to use `Read`, but actually not as performant. And this PR here is supposed to solve ... performance problems.

Reply:
I've just been profiling the whole patch. There's currently still one copy happening (the whole data buffer at once) in `SidecarServer::send_trace_v04`. That single copy accounts for 35% of the whole overhead. So copying here is most likely unacceptable.

Reply:
Yeah for sure we don't want to copy the strings we read, but for that you can make cleaner abstractions than using `read_string_ref_nomut`.


We could very well rewrite it to use the following primitives to encapsulate the unsafe manipulation in the Bytes implementation

```rust
impl Bytes {
  // make the bytes instance point to original_buffer[n..]
  // returns None if the buffer is smaller than n
  fn advance((&mut self, n: usize) -> Option<()> {
    if n > self.len { return None }
    self.ptr += n;
    self.len -= n;
    Ok(())
  }
  
  // make the bytes instance point to original_buffer[..n]
  // returns None if the buffer is smaller than n
  fn shrink((&mut self, n: usize) -> Option<()> {
    if n > self.len { return None }
    self.len = n;
    Ok(())
  }

  // Yield the first n bytes and advance the offset from as much 
  fn split_front_ref(&mut self, n: usize) -> Option<&[u8]> {
    let start = self.ptr;
    self.advance(n)?
    // Safe because the lifetime of the slice is smaller than the Bytes
    // and n smaller than length
    Ok(unsafe { slice:: from_raw_parts(start, n) })
  }
  
  fn split_front(&mut self, n: usize) -> Option<Bytes> {
    if n > self.len { return None }
    let mut front = self.clone();
    self.advance(n)?;
    front.shrink(n)?;
    Ok(front)
  }

}

fn read_string_ref(bytes: &'a mut Bytes) -> Result<&'a str, DecodeError> {
  let len = rpm::decode::read_str_len(bytes)?;
  let str_ref = self.split_front_ref(len).map_err(|| DecodeStringError::BufferSizeTooSmall(len))?;
  str::from_utf8(str_ref).map_err(|_| DecodeStringError::InvalidUtf8())
}

fn read_string_bytes(bytes: &'a mut Bytes) -> Result<&'a str, DecodeError> {
  let len = rpm::decode::read_str_len(bytes)?;
  let str_bytes = self.split_front(len).map_err(|| DecodeStringError::BufferSizeTooSmall(len))?;
  BytesString::from_bytes(str_bytes).map_err(...)
}
```

Reply:
By the way, I'm curious can your share the profile?

Reply:
I personally feel you're trying to introduce complexity where trivial unsafe is keeping it simple.

Also, fwiw - your read_string_ref does an extra utf8 check, so you'd use the unsafe method to convert to &str with _unchecked.

And we still rely on io::Read - which makes it easy to accidentally use it in contexts it's not supposed to be used (because it would copy).

Alternatively you could use something like:
```
struct RmpBuffer {
    pub bytes: Bytes,
    rmp: decode::Bytes<'static>,
}

impl RmpBuffer {
    pub fn new(mut bytes: Bytes) -> RmpBuffer {
        RmpBuffer {
            rmp: decode::Bytes::new(unsafe { bytes.as_mut_slice() }),
            bytes,
        }
    }
}
```
Basically the fast Bytes abstraction from Rmp itself. But it works on a slice, so you'll have to carry Bytes around to slice from it. But it would reduce the whole thing to a single unsafe {}.

Reply:
And sure, the profile looks like this: 
![image](https://github.com/user-attachments/assets/e42a4727-7f86-4b3e-b69b-36bb4e937f16)

Reply:
> Also, fwiw - your read_string_ref does an extra utf8 check, so you'd use the unsafe method to convert to &str with _unchecked.

It doesn't. `rpm::decode::read_str_len` doesn't do any validation and just read the length header. So when the  Bytes instance gets sliced, it's necessary to check that the slice you get is a valid string.

Reply:
Oh, my mistake.

Reply:
@bwoebi and @paullegranddc - Is this conversation still ongoing? Any changes either of you require here? I don't have a strong opinion, so happy to implement whatever the group agrees on.

Reply:
Feel free to implement @paullegranddc suggestion with io::Read and the additional methods to Bytes.

Reply:
On my end, I don't have a strong opinion. The code currently works, so I think we can merge it, but it'd be nice if you have time to refactor it. Either in this PR or latter

Reply:
@paullegranddc @bwoebi - I do like the `impl io::Read` solution as it seems to contain less sharp edges, but benchmarks wind up being 20% slower overall for the PR vs 10% faster with the `as_mut_slice` implementation. It's possible I misunderstood the suggestion and implemented something incorrectly here. But if neither of you see any glaring errors I will revert to the `as_mut_slice` implementation for now and create a ticket to follow up on this in another PR.

Reply:
This does not surprise me too much, at least it would need much more optimization by the compiler. So, if it's actually 30% slower, then please revert it.